andi/andi.py

```diff
@@ -1,62 +1,302 @@
 # -*- coding: utf-8 -*-
+from collections import OrderedDict
 from typing import (
     Dict, List, Optional, Type, Callable, Union, Container,
-    get_type_hints,
-)
+    get_type_hints, Tuple, cast, MutableMapping)
 
-from andi.typeutils import get_union_args, is_union, get_globalns
+from andi.typeutils import (
+    get_union_args,
+    is_union,
+    get_globalns,
+    get_unannotated_params,
+)
+from andi.errors import CyclicDependencyError, NonProvidableError
 
 
 def inspect(func: Callable) -> Dict[str, List[Optional[Type]]]:
     """
     For each argument of the ``func`` return a list of possible types.
+    Non annotated arguments are also returned with an empty list of possible
+    types.
     """
     globalns = get_globalns(func)
     annotations = get_type_hints(func, globalns)
+    for name in get_unannotated_params(func, annotations):
+        annotations[name] = None
     annotations.pop('return', None)
-    annotations.pop('cls', None)  # FIXME: pop first argument of methods
+    annotations.pop('self', None)  # FIXME: pop first argument of methods
+    annotations.pop('cls', None)
     res = {}
     for key, tp in annotations.items():
         if is_union(tp):
             res[key] = get_union_args(tp)
         else:
-            res[key] = [tp]
+            res[key] = [] if tp is None else [tp]
     return res
 
 
-def to_provide(
-        arguments_or_func: Union[
-            Callable,
-            Dict[str, List[Optional[Type]]]
-        ],
-        can_provide: Union[
-            Container[Type],
-            Callable[[Optional[Type]], bool]
-        ]
-        ) -> Dict[str, Optional[Type]]:
+ContainerOrCallableType = Union[
+    Container[Callable],
+    Callable[[Callable], bool]
+]
+
+
+PlanList = List[Tuple[Callable, Dict[str, Callable]]]
```

Comment:
Naming again :) I'd expect PlanList to be List[Plan], not a base class for Plan.

Idea, on how to solve it: make this type internal; return Plan(self[-1]) from plan.dependencies; expose only Plan type. It doesn't cover explicit `plan[2:4]` though.

Reply:
Maybe is just a naming problem. What if it is renamed to `StepList` or `PlanStepList`?. I think it would make sense:
* A plan is a list of steps (so it makes sense to inherit from it)
* Plan.dependencies is also a list of steps. 

> Idea, on how to solve it: make this type internal; return Plan(self[-1]) from plan.dependencies; expose only Plan type. It doesn't cover explicit plan[2:4] though.

I don't get this idea. Do you mean:

```py
_PlanList = List[Tuple[Callable, Dict[str, Callable]]] 

class Plan(_PlanList):

  @property 
  def dependencies(self) -> Plan:
    return Plan(self[:-1])

  # ...
```

But in this case it would be possible `plan.dependencies.dependencies` and `plan.dependencies.final_arguments`. I'm not sure this is desirable, but by the other side is not so harmful. So I don't know, I have no preference here.

Reply:
Meanwhile, we decided, I have changed the name to `PlanStepList` provisionally here: https://github.com/scrapinghub/andi/pull/8/commits/cddc84f9db0379fdb44a80fa1c0fb06cb43866bf

`PlanList` is clearly confusing